File: java/lsp/src/test/java/dev/mutwakil/androidide/lsp/java/providers/KeywordCompletionTest.java
package dev.mutwakil.androidide.lsp.java.providers;

import static com.google.common.truth.Truth.assertThat;

import dev.mutwakil.androidide.lsp.java.JavaLSPTest;
import dev.mutwakil.androidide.lsp.models.CompletionParams;
import dev.mutwakil.androidide.models.Position;
import dev.mutwakil.androidide.progress.ICancelChecker;
import java.util.List;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

/**
 * Tests keyword completion in different Java 17 syntactic contexts.
 *
 * <p>These tests are intentionally broad for now. The main purpose is to run completion at
 * carefully selected AST positions so that KeywordCompletionProvider can log and expose the
 * actual TreePath for each context.
 *
 * @author Akash Yadav
 */
@RunWith(RobolectricTestRunner.class)
public class KeywordCompletionTest {

  @Before
  public void setup() {
    JavaLSPTest.setup();
  }

  @Test
  public void packageDeclaration() {
    probe("completion/keywords/PackageDeclarationTest");
  }

  @Test
  public void importDeclaration() {
    probe("completion/keywords/ImportDeclarationTest");
  }

  @Test
  public void topLevelTypeDeclaration() {
    probe("completion/keywords/TopLevelTypeDeclarationTest");
  }

  @Test
  public void classModifiers() {
    probe("completion/keywords/ClassModifiersTest");
  }

  @Test
  public void classBody() {
    probe("completion/keywords/ClassBodyTest");
  }

  @Test
  public void fieldModifiers() {
    probe("completion/keywords/FieldModifiersTest");
  }

  @Test
  public void fieldType() {
    probe("completion/keywords/FieldTypeTest");
  }

  @Test
  public void methodModifiers() {
    probe("completion/keywords/MethodModifiersTest");
  }

  @Test
  public void methodReturnType() {
    probe("completion/keywords/MethodReturnTypeTest");
  }

  @Test
  public void methodThrows() {
    probe("completion/keywords/MethodThrowsTest");
  }

  @Test
  public void methodBody() {
    probe("completion/keywords/MethodBodyTest");
  }

  @Test
  public void localVariableType() {
    probe("completion/keywords/LocalVariableTypeTest");
  }

  @Test
  public void localVariableModifiers() {
    probe("completion/keywords/LocalVariableModifiersTest");
  }

  @Test
  public void expression() {
    probe("completion/keywords/ExpressionTest");
  }

  @Test
  public void throwStatement() {
    probe("completion/keywords/ThrowStatementTest");
  }

  @Test
  public void returnStatement() {
    probe("completion/keywords/ReturnStatementTest");
  }

  @Test
  public void ifStatement() {
    probe("completion/keywords/IfStatementTest");
  }

  @Test
  public void elseStatement() {
    probe("completion/keywords/ElseStatementTest");
  }

  @Test
  public void forStatement() {
    probe("completion/keywords/ForStatementTest");
  }

  @Test
  public void whileStatement() {
    probe("completion/keywords/WhileStatementTest");
  }

  @Test
  public void doStatement() {
    probe("completion/keywords/DoStatementTest");
  }

  @Test
  public void switchCase() {
    probe("completion/keywords/SwitchCaseTest");
  }

  @Test
  public void switchDefault() {
    probe("completion/keywords/SwitchDefaultTest");
  }

  @Test
  public void switchExpressionYield() {
    probe("completion/keywords/SwitchExpressionYieldTest");
  }

  @Test
  public void tryStatement() {
    probe("completion/keywords/TryStatementTest");
  }

  @Test
  public void catchClause() {
    probe("completion/keywords/CatchClauseTest");
  }

  @Test
  public void finallyClause() {
    probe("completion/keywords/FinallyClauseTest");
  }

  @Test
  public void synchronizedStatement() {
    probe("completion/keywords/SynchronizedStatementTest");
  }

  @Test
  public void tryWithResources() {
    probe("completion/keywords/TryWithResourcesTest");
  }

  @Test
  public void constructorBody() {
    probe("completion/keywords/ConstructorBodyTest");
  }

  @Test
  public void constructorInvocation() {
    probe("completion/keywords/ConstructorInvocationTest");
  }

  @Test
  public void lambdaBody() {
    probe("completion/keywords/LambdaBodyTest");
  }

  @Test
  public void classExtends() {
    probe("completion/keywords/ClassExtendsTest");
  }

  @Test
  public void classImplements() {
    probe("completion/keywords/ClassImplementsTest");
  }

  @Test
  public void interfaceBody() {
    probe("completion/keywords/InterfaceBodyTest");
  }

  @Test
  public void interfaceMethod() {
    probe("completion/keywords/InterfaceMethodTest");
  }

  @Test
  public void annotationInterfaceBody() {
    probe("completion/keywords/AnnotationInterfaceBodyTest");
  }

  @Test
  public void annotationDefault() {
    probe("completion/keywords/AnnotationDefaultTest");
  }

  @Test
  public void enumDeclaration() {
    probe("completion/keywords/EnumDeclarationTest");
  }

  @Test
  public void enumConstants() {
    probe("completion/keywords/EnumConstantsTest");
  }

  @Test
  public void enumBody() {
    probe("completion/keywords/EnumBodyTest");
  }

  @Test
  public void enumConstantClassBody() {
    probe("completion/keywords/EnumConstantClassBodyTest");
  }

  @Test
  public void recordDeclaration() {
    probe("completion/keywords/RecordDeclarationTest");
  }

  @Test
  public void recordHeader() {
    probe("completion/keywords/RecordHeaderTest");
  }

  @Test
  public void recordBody() {
    probe("completion/keywords/RecordBodyTest");
  }

  @Test
  public void recordConstructor() {
    probe("completion/keywords/RecordConstructorTest");
  }

  @Test
  public void sealedModifier() {
    probe("completion/keywords/SealedModifierTest");
  }

  @Test
  public void nonSealedModifier() {
    probe("completion/keywords/NonSealedModifierTest");
  }

  @Test
  public void permitsClause() {
    probe("completion/keywords/PermitsClauseTest");
  }

  @Test
  public void primitiveType() {
    probe("completion/keywords/PrimitiveTypeTest");
  }

  @Test
  public void newExpression() {
    probe("completion/keywords/NewExpressionTest");
  }

  @Test
  public void thisExpression() {
    probe("completion/keywords/ThisExpressionTest");
  }

  @Test
  public void superExpression() {
    probe("completion/keywords/SuperExpressionTest");
  }

  @Test
  public void instanceofExpression() {
    probe("completion/keywords/InstanceofExpressionTest");
  }

  private void probe(String file) {
    JavaLSPTest.apply(
        () -> {
          openFile(file);

          Position position = cursorPosition();

          List<CharSequence> items = completionTitles(position);

          System.out.println();
          System.out.println("==================================================");
          System.out.println("Keyword completion probe: " + file);
          System.out.println("Cursor: " + position);
          System.out.println("Items: " + items);
          System.out.println("==================================================");
          System.out.println();

          assertThat(items).isNotNull();
        });
  }

  private List<CharSequence> completionTitles(Position position) {
    return JavaLSPTest.server
        .complete(
            new CompletionParams(
                position,
                JavaLSPTest.file,
                ICancelChecker.NOOP) {
              {
                setPrefix("");
              }
            })
        .getItems()
        .stream()
        .map(item -> item.getIdeLabel())
        .collect(java.util.stream.Collectors.toList());
  }
}

File: java/lsp/src/test/java/dev/mutwakil/androidide/lsp/java/providers/KeywordCompletionTest.kt
 /*
  *  This file is part of AndroidIDE.
  *
  *  AndroidIDE is free software: you can redistribute it and/or modify
  *  it under the terms of the GNU General Public License as published by
  *  the Free Software Foundation, either version 3 of the License, or
  *  (at your option) any later version.
  *
  *  AndroidIDE is distributed in the hope that it will be useful,
  *  but WITHOUT ANY WARRANTY; without even the implied warranty
  *  of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  *  GNU General Public License for more details.
  *
  *  You should have received a copy of the GNU General Public License
  *   along with AndroidIDE.  If not, see <https://www.gnu.org/licenses/>.
  */

package dev.mutwakil.androidide.lsp.java.providers

import com.google.common.truth.Truth.assertThat
import dev.mutwakil.androidide.lsp.java.JavaLSPTest
import dev.mutwakil.androidide.lsp.models.CompletionParams
import dev.mutwakil.androidide.models.Position
import dev.mutwakil.androidide.progress.ICancelChecker
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class KeywordCompletionTest {

  @Before
  fun setup() {
    JavaLSPTest.setup()
  }

  @Test
  fun packageDeclaration() {
    probe("completion/keywords/PackageDeclarationTest")
  }

  @Test
  fun importDeclaration() {
    probe("completion/keywords/ImportDeclarationTest")
  }

  @Test
  fun topLevelTypeDeclaration() {
    probe("completion/keywords/TopLevelTypeDeclarationTest")
  }

  @Test
  fun classModifiers() {
    probe("completion/keywords/ClassModifiersTest")
  }

  @Test
  fun classBody() {
    probe("completion/keywords/ClassBodyTest")
  }

  @Test
  fun fieldModifiers() {
    probe("completion/keywords/FieldModifiersTest")
  }

  @Test
  fun fieldType() {
    probe("completion/keywords/FieldTypeTest")
  }

  @Test
  fun methodModifiers() {
    probe("completion/keywords/MethodModifiersTest")
  }

  @Test
  fun methodReturnType() {
    probe("completion/keywords/MethodReturnTypeTest")
  }

  @Test
  fun methodThrows() {
    probe("completion/keywords/MethodThrowsTest")
  }

  @Test
  fun methodBody() {
    probe("completion/keywords/MethodBodyTest")
  }

  @Test
  fun localVariableType() {
    probe("completion/keywords/LocalVariableTypeTest")
  }

  @Test
  fun localVariableModifiers() {
    probe("completion/keywords/LocalVariableModifiersTest")
  }

  @Test
  fun expression() {
    probe("completion/keywords/ExpressionTest")
  }

  @Test
  fun throwStatement() {
    probe("completion/keywords/ThrowStatementTest")
  }

  @Test
  fun returnStatement() {
    probe("completion/keywords/ReturnStatementTest")
  }

  @Test
  fun ifStatement() {
    probe("completion/keywords/IfStatementTest")
  }

  @Test
  fun elseStatement() {
    probe("completion/keywords/ElseStatementTest")
  }

  @Test
  fun forStatement() {
    probe("completion/keywords/ForStatementTest")
  }

  @Test
  fun whileStatement() {
    probe("completion/keywords/WhileStatementTest")
  }

  @Test
  fun doStatement() {
    probe("completion/keywords/DoStatementTest")
  }

  @Test
  fun switchCase() {
    probe("completion/keywords/SwitchCaseTest")
  }

  @Test
  fun switchDefault() {
    probe("completion/keywords/SwitchDefaultTest")
  }

  @Test
  fun switchExpressionYield() {
    probe("completion/keywords/SwitchExpressionYieldTest")
  }

  @Test
  fun tryStatement() {
    probe("completion/keywords/TryStatementTest")
  }

  @Test
  fun catchClause() {
    probe("completion/keywords/CatchClauseTest")
  }

  @Test
  fun finallyClause() {
    probe("completion/keywords/FinallyClauseTest")
  }

  @Test
  fun synchronizedStatement() {
    probe("completion/keywords/SynchronizedStatementTest")
  }

  @Test
  fun tryWithResources() {
    probe("completion/keywords/TryWithResourcesTest")
  }

  @Test
  fun constructorBody() {
    probe("completion/keywords/ConstructorBodyTest")
  }

  @Test
  fun constructorInvocation() {
    probe("completion/keywords/ConstructorInvocationTest")
  }

  @Test
  fun lambdaBody() {
    probe("completion/keywords/LambdaBodyTest")
  }

  @Test
  fun classExtends() {
    probe("completion/keywords/ClassExtendsTest")
  }

  @Test
  fun classImplements() {
    probe("completion/keywords/ClassImplementsTest")
  }

  @Test
  fun interfaceBody() {
    probe("completion/keywords/InterfaceBodyTest")
  }

  @Test
  fun interfaceMethod() {
    probe("completion/keywords/InterfaceMethodTest")
  }

  @Test
  fun annotationInterfaceBody() {
    probe("completion/keywords/AnnotationInterfaceBodyTest")
  }

  @Test
  fun annotationDefault() {
    probe("completion/keywords/AnnotationDefaultTest")
  }

  @Test
  fun enumDeclaration() {
    probe("completion/keywords/EnumDeclarationTest")
  }

  @Test
  fun enumConstants() {
    probe("completion/keywords/EnumConstantsTest")
  }

  @Test
  fun enumBody() {
    probe("completion/keywords/EnumBodyTest")
  }

  @Test
  fun enumConstantClassBody() {
    probe("completion/keywords/EnumConstantClassBodyTest")
  }

  @Test
  fun recordDeclaration() {
    probe("completion/keywords/RecordDeclarationTest")
  }

  @Test
  fun recordHeader() {
    probe("completion/keywords/RecordHeaderTest")
  }

  @Test
  fun recordBody() {
    probe("completion/keywords/RecordBodyTest")
  }

  @Test
  fun recordConstructor() {
    probe("completion/keywords/RecordConstructorTest")
  }

  @Test
  fun sealedModifier() {
    probe("completion/keywords/SealedModifierTest")
  }

  @Test
  fun nonSealedModifier() {
    probe("completion/keywords/NonSealedModifierTest")
  }

  @Test
  fun permitsClause() {
    probe("completion/keywords/PermitsClauseTest")
  }

  @Test
  fun primitiveType() {
    probe("completion/keywords/PrimitiveTypeTest")
  }

  @Test
  fun newExpression() {
    probe("completion/keywords/NewExpressionTest")
  }

  @Test
  fun thisExpression() {
    probe("completion/keywords/ThisExpressionTest")
  }

  @Test
  fun superExpression() {
    probe("completion/keywords/SuperExpressionTest")
  }

  @Test
  fun instanceofExpression() {
    probe("completion/keywords/InstanceofExpressionTest")
  }

  private fun probe(file: String) {
    JavaLSPTest.apply {
      openFile(file)

      val pos = cursorPosition()
      val items = completionTitles(pos)

      println()
      println("==================================================")
      println("Keyword completion probe: $file")
      println("Cursor: $pos")
      println("Items:")
      items.forEach { println("  - $it") }
      println("==================================================")
      println()

      assertThat(items).isNotNull()
    }
  }

  private fun completionTitles(pos: Position): List<CharSequence> {
    return JavaLSPTest.server
      .complete(
        CompletionParams(
          pos,
          JavaLSPTest.file!!,
          ICancelChecker.NOOP,
        ).apply {
          prefix = ""
        },
      )
      .items
      .map { it.ideLabel }
  }
}